File: app/src/main/java/com/tommasov/mg4swipenovalauncher/MainActivity.java
package com.tommasov.mg4swipenovalauncher;

import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.widget.ListView;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import android.widget.Button;

public class MainActivity extends AppCompatActivity {
    private PackageManager packageManager;
    private AppListAdapter adapter;
    private List<ApplicationInfo> allApps;
    private PreferencesManager preferencesManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_main);

        preferencesManager = new PreferencesManager(this);

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        TextView explanationText = findViewById(R.id.explanation_text);
        explanationText.setText(R.string.explanation_text);

        String currentPackageName = getPackageName();
        packageManager = getPackageManager();
        List<ApplicationInfo> userApps = new ArrayList<>();

        for (ApplicationInfo appInfo : packageManager.getInstalledApplications(PackageManager.GET_META_DATA)) {
            if (!appInfo.packageName.equals(currentPackageName)) {
                userApps.add(appInfo);
            }
        }

        Collections.sort(userApps, (app1, app2) -> {
            String label1 = app1.loadLabel(packageManager).toString();
            String label2 = app2.loadLabel(packageManager).toString();
            return label1.compareToIgnoreCase(label2);
        });

        String selectedPackage = preferencesManager.getSelectedPackage();
        if (selectedPackage == null) {
            for (ApplicationInfo appInfo : userApps) {
                if (appInfo.packageName.equals("com.teslacoilsw.launcher")) {
                    selectedPackage = appInfo.packageName;
                    preferencesManager.saveSelectedPackage(selectedPackage);
                    break;
                }
            }
        }

        ListView listView = findViewById(R.id.app_list);
        adapter = new AppListAdapter(this, userApps, selectedPackage);
        listView.setAdapter(adapter);

        listView.setOnItemClickListener((parent, view, position, id) -> {
            ApplicationInfo selectedApp = userApps.get(position);
            preferencesManager.saveSelectedPackage(selectedApp.packageName);
            adapter.setSelectedPackage(selectedApp.packageName);
            adapter.notifyDataSetChanged();
            Toast.makeText(MainActivity.this, "Selected: " + selectedApp.packageName, Toast.LENGTH_SHORT).show();
        });

        Button toggleSystemAppsButton = findViewById(R.id.toggle_system_apps_button);
        toggleSystemAppsButton.setOnClickListener(v -> {
            adapter.toggleSystemAppsVisibility();
            toggleSystemAppsButton.setText(adapter.isSystemAppsVisible() ? getString(R.string.hide_system_apps) : getString(R.string.show_system_apps));
        });

        Switch switchBackButton = findViewById(R.id.switch_back_button);

        switchBackButton.setChecked(!preferencesManager.getBackButtonVisibility().equals("VISIBLE"));

        switchBackButton.setOnClickListener(view -> {
            if (preferencesManager.getBackButtonVisibility().equals("VISIBLE")) {
                preferencesManager.saveBackButtonVisibility("INVISIBLE");
                switchBackButton.setChecked(true);
            } else {
                preferencesManager.saveBackButtonVisibility("VISIBLE");
                switchBackButton.setChecked(false);
            }
            stopSwipeService();
            startSwipeService();
        });

        startSwipeService();
    }

    private void stopSwipeService() {
        Intent intent = new Intent(this, SwipeService.class);
        stopService(intent);
    }

    private void startSwipeService() {
        Intent intent = new Intent(this, SwipeService.class);
        startService(intent);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }
}

File: app/src/main/java/com/tommasov/mg4swipenovalauncher/MainActivity.kt
package com.tommasov.mg4swipenovalauncher

import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.tommasov.mg4swipenovalauncher.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var prefs: PreferencesManager
    private lateinit var adapter: AppListAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        prefs = PreferencesManager.getInstance(this)

        ViewCompat.setOnApplyWindowInsetsListener(binding.main) { v, insets ->
            val bars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(bars.left, bars.top, bars.right, bars.bottom)
            insets
        }

        binding.explanationText.setText(R.string.explanation_text)

        val pm = packageManager
        val userApps = pm.getInstalledApplications(PackageManager.GET_META_DATA)
            .filter { it.packageName != packageName }
            .sortedBy { it.loadLabel(pm).toString().lowercase() }

        var selectedPackage = prefs.selectedPackage
        if (selectedPackage == null) {
            userApps.find { it.packageName == SwipeService.DEFAULT_LAUNCHER }?.let {
                selectedPackage = it.packageName
                prefs.selectedPackage = selectedPackage
            }
        }

        adapter = AppListAdapter(this, userApps, selectedPackage)
        binding.appList.adapter = adapter

        binding.appList.setOnItemClickListener { _, _, position, _ ->
            val app = userApps[position]
            prefs.selectedPackage = app.packageName
            adapter.setSelectedPackage(app.packageName)
            adapter.notifyDataSetChanged()
            Toast.makeText(this, String.format(getString(R.string.toast_selected), app.loadLabel(pm)), Toast.LENGTH_SHORT).show()
        }

        binding.toggleSystemAppsButton.setOnClickListener {
            adapter.toggleSystemAppsVisibility()
            binding.toggleSystemAppsButton.text = getString(
                if (adapter.isSystemAppsVisible()) R.string.hide_system_apps else R.string.show_system_apps
            )
        }

        binding.switchBackButton.isChecked = prefs.isBackButtonHidden
        binding.switchBackButton.setOnClickListener {
            val hide = !prefs.isBackButtonHidden
            prefs.backButtonVisibility = if (hide) "INVISIBLE" else "VISIBLE"
            binding.switchBackButton.isChecked = hide
            stopSwipeService()
            startSwipeService()
        }

        binding.uninstallButton.setOnClickListener {
            androidx.appcompat.app.AlertDialog.Builder(this)
                .setTitle(getString(R.string.uninstall))
                .setMessage(getString(R.string.uninstall_confirm))
                .setPositiveButton(android.R.string.ok) { _, _ ->
                    stopSwipeService()
                    val intent = Intent(Intent.ACTION_DELETE, Uri.parse("package:$packageName"))
                    startActivity(intent)
                }
                .setNegativeButton(android.R.string.cancel, null)
                .show()
        }

        startSwipeService()
    }

    private fun stopSwipeService() {
        try { stopService(Intent(this, SwipeService::class.java)) }
        catch (e: Exception) { AppLogger.e("Failed to stop SwipeService", e) }
    }

    private fun startSwipeService() {
        try { startForegroundService(Intent(this, SwipeService::class.java)) }
        catch (e: Exception) { AppLogger.e("Failed to start SwipeService", e) }
    }
}
